pkg/controller/controller_instance.go

```diff
@@ -329,6 +367,165 @@ func (c *controller) initOrphanMitigationCondition(instance *v1beta1.ServiceInst
 	return false, nil
 }
 
+// getInstanceAndGenerationAsKey returns a string with the format of
+// namespace/instancename@generation
+func (c *controller) getInstanceAndGenerationAsKey(instance *v1beta1.ServiceInstance) (string, error) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(instance)
+	if err != nil {
+		err := fmt.Errorf("Couldn't get key for object %+v: %v", instance, err)
+		return "", err
+	}
+
+	key = strings.Join([]string{key, "@", strconv.FormatInt(instance.Generation, 10)}, "")
+	return key, nil
+}
+
+// setRetryBackoffRequired marks the specified instance/generation as needing a
+// delay before the next provision/update is attempted.  We always set this flag
+// before attempting an update or delete operation in case we must retry.  This
+// will eventually be cleared by background worker running purgeExpiredRetryEntries()
+// or when the operation is successful.
+func (c *controller) setRetryBackoffRequired(instance *v1beta1.ServiceInstance) {
+	key, err := c.getInstanceAndGenerationAsKey(instance)
+	if err != nil {
+		glog.Errorf(err.Error())
+		return
+	}
+
+	c.instanceOperationRetryQueue.mutex.Lock()
+	defer c.instanceOperationRetryQueue.mutex.Unlock()
+	// older generations are meaningless, purge any
+	deleteInstanceKeysFromMapForAllGens(c.instanceOperationRetryQueue.backoffBeforeRetrying, key)
+	c.instanceOperationRetryQueue.backoffBeforeRetrying[key] = time.Now()
+	glog.V(4).Infof("added %v to backoffBeforeRetrying", key)
+}
+
+// deleteInstanceKeysFromMapForAllGens purges out any instance entries ignoring the generation
+// Assumes the caller did the neccessary mutex locking for thread safety
+func deleteInstanceKeysFromMapForAllGens(m map[string]time.Time, key string) {
+	end := strings.Index(key, "@")
+	if end > 1 {
+		anyGeneration := key[0:end]
+		for k := range m {
```

Comment:
Do we ever need to keep around the times from previous generations? Why don't we move the generation into the value part of the map entry instead of embedding it in the key part. That way you don't have to iterate over the entire map to clear out entries for old generations. When accessing the map, you would look up by namespace/name and then verify that the generation in the value part of the entry matches the current generation of the instance.

Reply:
We dont keep or care about previous generations, I like the suggestion.